src/lib/storage/buffer/frame.hpp

```diff
@@ -0,0 +1,173 @@
+ *                                                                                    ^                                     ^ unlock_shared   |    |
+ *                                                                                    | unlock_shared                       +-----------------+    |
+ *                                                                                    |                                                            |
+ *                                                                                    | (only if no other latches are present)                     |
+ *                                                                                    +------------------------------------------------------------+
+ */
+class Frame final : private Noncopyable {
+ public:
+  // Type of state and version
+  using StateVersionType = uint64_t;
+
+  // UNLOCKED has a value of 0
+  static constexpr StateVersionType UNLOCKED = 0;
+
+  // SINGLE_LOCKED_SHARED signifies that the frame has a single reader. It has a value of 1. Used for convenience.
+  static constexpr StateVersionType SINGLE_LOCKED_SHARED = 1;
+
+  // LOCKED_SHARED signifies that the frame has the maxium amount of shared readers. It has a value of 65532.
+  static constexpr StateVersionType MAX_LOCKED_SHARED = 0xFFFF - 3;
+
+  // LOCKED signifies that the frame is locked exclusively. It has a value of 65533.
+  static constexpr StateVersionType LOCKED = 0xFFFF - 2;
+
+  // MARKED signifies that the frame is marked for eviction. It has a value of 65534.
+  static constexpr StateVersionType MARKED = 0xFFFF - 1;
+
+  // EVICTED has a value of 65535 and is the initial state of the frame.
+  static constexpr StateVersionType EVICTED = 0xFFFF;
+
+  // Constructs the frame in state EVICTED and version 0.
+  Frame();
+
+  // Set the NUMA node of the frame.
+  void set_node_id(const NodeID node_id);
+
+  // Set the dirty flag of the frame. Fails if the frame is not locked exclusively.
+  void mark_dirty();
+
+  // Check if the frame is dirty.
+  bool is_dirty() const;
+
+  // Reset the dirty flag. Fails if the frame is not locked exclusively.
+  void reset_dirty();
+
+  // Get the NUMA node of the frame.
+  NodeID node_id() const;
+
+  /**
+   * Try to latch in frame in shared node by incrementing the state. Fails if the frame is 
+   * not in state UNLOCKED, MARKED or reaches MAX_LOCKED_SHARED.
+   */
+  bool try_lock_shared(const StateVersionType old_state_and_version);
+
+  // Try to latch in frame in exclusive node. Fails if the frame is not in state UNLOCKED or MARKED.
+  bool try_lock_exclusive(const StateVersionType old_state_and_version);
+
+  // Try to mark the frame. Fails if the frame is not in state UNLOCKED.
+  bool try_mark(const StateVersionType old_state_and_version);
+
+  // Decrement the shared latch. Reaches UNLOCKED if the state is 0.
```

Comment:
Similar to the cpp, mentioning when this function returns `true` would clarify the description. `Reaches UNLOCKED` is not clear imo